Add utility functions to get interface from identifier

ivanpauno opened the pull request:
To avoid repeated code in some `ros2cli` packages.
See https://github.com/ros2/ros2cli/pull/301#discussion_r310283612.

artivis commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,61 @@
+
+
+def full_type_to_namespaced_type(path: Text):
+    """Create a `NamespacedType` object from the full name of an action."""
```
```suggestion
    """Create a `NamespacedType` object from the full name of an interface."""
```

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,61 @@
+
+def is_msg(msg):
+    """Return `True` if `msg` is a message."""
+    return hasattr(msg, 'get_fields_and_field_types')
```
This is the old API from a message. Instead this should probably check for `SLOT_TYPES`.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,61 @@
+    return NamespacedType(namespace, name)
+
+
+def full_type_to_namespaced_type(path: Text):
```
Why should all the functions start with `full_`? Maybe renamed this function to `get_namespaced_type` and the following ones `get_message|service|action_namespaced_type`.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,61 @@
+
+def is_action(action):
+    """Return `True` if `action` is an action."""
+    return hasattr(action, 'Goal')
```
This should check all the attributes which are present in an action. I would also name the argument `interface` since it is unclear if it actually is an action.

Same for services below.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,61 @@
+    return hasattr(action, 'Goal')
+
+
+def is_srv(srv):
```
No abbreviation: `is_service`.

Same below for `is_msg`.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,61 @@
+    namespace = namespace[0:-1]
+    if len(namespace) == 1 and default_namespace is not None:
+        namespace.append(default_namespace)
+    return NamespacedType(namespace, name)
```
Please move this internal helper below all the functions using it.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def is_action(interface):
+    """Return `True` if `action` is an action."""
```
Docblock should mention `interface` instead.

Same below.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,65 @@
+
+def get_service_namespaced_type(path: Text):
+    """Create a `NamespacedType` object from the full name of a service."""
+    return __get_namespaced_type(path, 'srv')
```
Nitpick: I would use the following order bases on increasingly higher concepts:
* `get_message_namespaced_type`
* `get_service_namespaced_type`
* `get_action_namespaced_type`

Same for the `is_*` functions below.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,65 @@
+    return hasattr(interface, 'SLOT_TYPES')
+
+
+def __get_namespaced_type(path: Text, default_namespace=None):
```
This should have a single underscore prefix. There is no special semantic for double underscore outside of a class context.

ivanpauno commented:
@dirk-thomas Addressed your comments in https://github.com/ros2/rosidl_python/pull/78/commits/0f68df663a9e3d789701996e8d59c17be02e35ba.
I also added: `get_interface`, `get_message`, `get_service`, `get_action` in https://github.com/ros2/rosidl_python/pull/78/commits/55b011dd857ab0f06a282f8532d82e1686d53dd0, which I realized they will be usefull.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,95 @@
+    """Get a message from its full name."""
+    interface = import_message_from_namespaced_type(get_message_namespaced_type(path))
+    if not is_message(interface):
+        raise RuntimeError('Expected the full name of a message')
```
Maybe raise `ValueError` instead?

Also it would be good if the error message contains the argument within quotes.

dirk-thomas commented on rosidl_runtime_py/rosidl_runtime_py/utilities.py:
```diff
@@ -0,0 +1,95 @@
+
+def get_message_namespaced_type(path: Text):
+    """Create a `NamespacedType` object from the full name of a message."""
+    return _get_namespaced_type(path, 'msg')
```
Keyword arguments should be passed with the explicitly specified keyword: `default_namespace='msg'`.

ivanpauno commented:
CI (up to rosidl_generator
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7807)](http://ci.ros2.org/job/ci_linux/7807/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3885)](http://ci.ros2.org/job/ci_linux-aarch64/3885/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6373)](http://ci.ros2.org/job/ci_osx/6373/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7641)](http://ci.ros2.org/job/ci_windows/7641/)
